shm_wrapper/shm_wrapper.c

```diff
@@ -812,15 +833,16 @@ int gamepad_write(uint32_t pressed_buttons, float joystick_vals[4]) {
     my_sem_post(rd_sem, "robot_desc_mutex");
 
     // wait on gp_sem
-    my_sem_wait(gp_sem, "gamepad_mutex");
+    my_sem_wait(input_sem, "input_mutex");
 
-    gp_shm_ptr->buttons = pressed_buttons;
+    int index = source - GAMEPAD;  // Assumes KEYBOARD is after GAMEPAD in robot_desc_field_t enum
+    input_shm_ptr->inputs[index].buttons = pressed_buttons;
```

Comment:
I wonder if we can add a check here to make sure that they aren't any bits on at higher positions than the number of buttons on the input source (for example from passing in a button bitmap intended for a keyboard as a gamepad instead--gamepads don't have 47 buttons)

Reply:
Theres no need for a check as we don't read the higher bits anyways in the student API. Whoever is sending the data is responsible for checking the bitmap in your example, not Runtime.